src/main/java/com/redhat/jenkins/plugins/cachet/pipeline/CachetGatingWaitForResourcesStep.java

```diff
@@ -0,0 +1,163 @@
+
+                CachetGatingMetrics tmpResourceMet = metrics.get(resource);
+                if(tmpResourceMet.getGatingStatus().equals("Operational")){
+                    waitForResourcesMet.put(resource, tmpResourceMet.getGatedTimeElapsed());
+                }
+                else {
+                    resourcesToWaitFor.add(resource);
+                }
+            }
+            
+            while (!resourcesToWaitFor.isEmpty()){// So there are resources we need to wait for
+                sleepingHandler(resourcesToWaitFor);
+                for (String resource : resourcesToWaitFor){
+                    CachetGatingMetrics tmpResourceMet = getCurrentMetrics().get(resource);
+                    if(tmpResourceMet.getGatingStatus().equals("Operational")){
+                        resourcesToWaitFor.remove(resource);
+                        waitForResourcesMet.put(resource, tmpResourceMet.getGatedTimeElapsed());
+                    }
+                }
+            }
+          return waitForResourcesMet;
+        }
+
+        /**
+         * Help method
+         * @return The current metrics
+         * @throws Exception
+         */
+        public Map<String, CachetGatingMetrics> getCurrentMetrics() throws Exception {
+            Run run = getContext().get(Run.class);
+            Action action = run.getAction(com.redhat.jenkins.plugins.cachet.CachetGatingAction.class);
+            if (action == null) {
+                action = new CachetGatingAction();
+            }
+            return ((CachetGatingAction) action).getGatingMetricsMap();
+        }
+
+        /**
+         * Handle the waiting for the resources
+         * @param listener
+         * @param resourcesToWaitFor
+         * @throws Exception
+         */
+        public void sleepingHandler(List<String> resourcesToWaitFor) throws Exception {
+            TaskListener listener = getContext().get(TaskListener.class);
+            listener.getLogger().println("Waiting for " + resourcesToWaitFor.toString() + " for " + recurrencePeriod + " ms");
+            Thread.sleep(recurrencePeriod);
+            recurrencePeriod = Math.min((long)(recurrencePeriod * RECURRENCE_PERIOD_BACKOFF), MAX_RECURRENCE_PERIOD);
+        }
+    }
+
+    /**
+     * Adds the step as a workflow extension.
+     */
+    @Extension
+    public static class DescriptorImpl extends StepDescriptor {
+
+        @Override
+        public String getFunctionName() {
+            return "cachetWaitForResources";
+        }
+
+        @Override
+        public String getDisplayName() {
+            return "Cachet Gating Check Resources Status";
+        }
+
+        @Override
+        public boolean takesImplicitBlockArgument() {
+            return true;
+        }
+
+        @Override
+        public Set<? extends Class<?>> getRequiredContext() {
+            return ImmutableSet.of(Run.class, Launcher.class, TaskListener.class);
```

Comment:
I am speculating here, but is not requiring `Launcher` (you do not seem to be using) going to force users to wait for service inside a `node` step? That sounds wasteful to resources the node occupies. The suggested workflow should be the other way around IMO - wait for services to be ready and only then allocate resources.